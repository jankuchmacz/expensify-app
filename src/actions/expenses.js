import uuid from 'uuid';
import database from '../firebase/firebase';
//Action that we need
//Add-expense

//component calls action generator
//action generator returns function
//component dispatches function
//functions runs (has the ability to dispatch other actions and do whatever it wants)
export const addExpense = (expense)=>{   
    return {
        type: 'ADD_EXPENSE',
        expense: expense
    }
}
export const startAddExpense = (expnseData = {})=>{
    return (dispatch) => {
        //destructuring
        const {
            description = '', note='', amount = 0, createdAt = 0
        } = expnseData;
        const expense={
            description,
            note,
            amount,
            createdAt
        }
        return database.ref('expenses').push(expense).then((ref)=>{
           dispatch(addExpense({
                id: ref.key,
                ...expense
            })); 
        })
    };
}
//Remove expense
export const removeExpense = (id) => {
    return { 
        type: 'REMOVE_EXPENSE',
        id: id
    }
}
export const startRemoveExpense = ({id} = {}) =>{
    return (dispatch)=>{
        return database.ref(`expenses/${id}`).remove().then(()=>{
            dispatch(removeExpense(id))
        }).catch((e)=>{
            console.log('Error during fetching data from DB', e);
        })
    }
}

//Edit expense
export const editExpense = (id, updates) => {
    return {
        type: 'EDIT_EXPENSE',
        id,
        updates
    }
}
//SET_EXPENSES
export const setExpenses = (expenses)=>({
    type: 'SET_EXPENSES',
    expenses: expenses
});
export const startSetExpenses = () =>{
    return (dispatch)=>{
        return database.ref('expenses').once('value').then((snapashot)=>{
            const expenses=[];
            snapashot.forEach((childSnapshot)=>{
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            })
            dispatch(setExpenses(expenses));
        }).catch((e)=>{
            console.log('Error during fetching data from DB', e);
        })
    }
};
