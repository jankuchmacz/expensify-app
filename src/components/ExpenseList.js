import React from "react";
import {connect} from 'react-redux'; //connect connects component to redux store
import ExpenseListItem from "./ExpenseLisItem";
import getVisibleExpenses from "../selectors/expenses";

export const ExpenseList = (props) => {
    return (
        <div className="content-container">
            <div className="list-header">
                <div className="show-for-mobile">Expenses</div>
                <div className="show-for-desktop">Expense</div>
                <div className="show-for-desktop">Amount</div>
            </div>
            <div className="list-body">
            {
                props.expenses.length === 0 ? (
                    <div className="list-item list-item--message">
                        <span>No expenses</span>
                    </div>
                    
                ): (
                    props.expenses.map((expense) => {
                        return (
                            <ExpenseListItem key={expense.id} expense={expense} />
                        );            
                    })
                )
            }
            </div>
        </div>
    );
}
const mapStateToProps = (state) => {
    //in this function we determine what information do we need from the store
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    }
}
export default connect(mapStateToProps) (ExpenseList);
//if you connect component to redux store it becomes reactive
//as the store changes component is rerendered with new values
