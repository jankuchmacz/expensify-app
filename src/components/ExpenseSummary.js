import React from "react";
import {connect} from 'react-redux'; //connect connects component to redux store
import getExpensesTotal from "../selectors/expenses-total";
import getVisibleExpenses from "../selectors/expenses";
import numeral from "numeral";
import { Link } from "react-router-dom";
export const ExpenseSummary = (props) => {
    const numberOfExpenses = props.expenses;
    const expense = numberOfExpenses === 1 ? 'expense' : 'expenses'
    return (
        <div className='page-header'>
            <div className='content-container'>
                <h1 className='page-header__title'>
                    Viewing <span> {numberOfExpenses} </span> {expense} totalling <span>{numeral(props.expensesSum/100).format('$,0.00')}</span>
                </h1>
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add Expense</Link>
                </div>  
            </div>             
        </div>
    );
}
const mapStateToProps = (state) => {
    //in this function we determine what information do we need from the store
    const expenses = getVisibleExpenses(state.expenses, state.filters);
    return {
        expenses: expenses.length,
        expensesSum: getExpensesTotal(expenses)
    }
}
export default connect(mapStateToProps) (ExpenseSummary);
//if you connect component to redux store it becomes reactive
//as the store changes component is rerendered with new values