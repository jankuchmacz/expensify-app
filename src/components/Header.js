import React from 'react';
import { NavLink} from 'react-router-dom';
const Header = () => (
    <header>
        <h1>Expensify</h1>
        <NavLink to="/" activeClassName="is-active" exact={true}>Home</NavLink>
        <NavLink to="/create" activeClassName="is-active">New expense</NavLink>
        <NavLink to="/help" activeClassName="is-active">Help</NavLink>
    </header>
);
//we want to render header on all pages
//activeClassName - gives us possibility to apply style to currently clicked link (if we are on this page)

export default Header;