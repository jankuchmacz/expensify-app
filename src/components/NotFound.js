import React from "react";
import {Link} from 'react-router-dom';
const NotFoundPage = () => {
    return (
        <div>
            404! - <Link to="/">Go home</Link>
        </div>
    );
    //thanks to Link we are using client side routing not server side routng
    //due to that we avoid full page refresh
}
export default NotFoundPage;