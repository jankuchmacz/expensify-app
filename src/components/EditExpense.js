import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense } from "../actions/expenses";
import { removeExpense } from "../actions/expenses";

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');       
    }
    onClick = () => {
        this.props.removeExpense({id: this.props.expense.id})
        this.props.history.push('/');
    }
    render(){
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense} 
                    onSubmit = {this.onSubmit}
                />                
               <button onClick = {this.onClick}>
                    Remove
                </button>
            </div>
        );
    }
}
const mapDispatchToProps = (dispatch, props) => ({
    editExpense: (id, expense) => {
        return dispatch(editExpense(id, expense))
    },
    removeExpense: ({id}) => {
        return dispatch(removeExpense({id}))
    }
})
const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id
        })
    }
}
export default connect(mapStateToProps, mapDispatchToProps) (EditExpensePage);