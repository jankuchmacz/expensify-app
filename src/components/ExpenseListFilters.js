import React from "react";
import {connect} from 'react-redux';
import { setTextFilter } from "../actions/filters";
import { sortByAmount } from "../actions/filters";
import { sortByDate } from "../actions/filters";
import { DateRangePicker } from "react-dates";
import { setStartDate } from "../actions/filters";
import { setEndDate } from "../actions/filters";

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };
    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }
    onFocusChange = (focusedInput) => {
        this.setState(()=>({
            calendarFocused: focusedInput
        }))
        //console.log(focusedInput);
    }
    onTextChange = (e) => {
        //every time input changes this function fires
        this.props.setTextFilter(e.target.value);
        //console.log(e.target.value);
    }
    onSortChange = (e) => {
        const sortBy = e.target.value;
        if(sortBy==='amount') {
            this.props.sortByAmount();
        } else if (sortBy==='date'){
            this.props.sortByDate()
        }
    }
    render () {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input 
                            type="text" 
                            className="text-input"
                            placeholder="Search expenses"
                            value={this.props.filters.text} 
                            onChange={this.onTextChange} 
                        />
                    </div>
                    <div className="input-group__item">
                        <select 
                            className="select"
                            value={this.props.filters.sortBy} 
                            onChange={this.onSortChange}>
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker 
                            startDate={this.props.filters.startDate}
                            endDate = {this.props.filters.endDate}
                            onDatesChange = {this.onDatesChange}
                            focusedInput= {this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            showClearDates={true}
                            numberOfMonths = {1}
                            isOutsideRange = {() => false}
                            startDateId="start"
                            endDateId="end"
                        />
                    </div>
                </div>
            </div>
        );
    }
}
const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}
const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => {
        return dispatch(setTextFilter(text));
    },
    sortByDate: () => {
        return dispatch(sortByDate());
    },
    sortByAmount: () => {
        return dispatch(sortByAmount());
    },
    setStartDate: (startDate) => {
        return dispatch(setStartDate(startDate));
    },
    setEndDate: (endDate) => {
        return dispatch(setEndDate(endDate));
    }
})
export default connect(mapStateToProps, mapDispatchToProps) (ExpenseListFilters);