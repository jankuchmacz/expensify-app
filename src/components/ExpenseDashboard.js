import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummary from './ExpenseSummary';

const ExpenseDashoboardPage = () => (
    <div>
        <ExpenseSummary/>
        <ExpenseListFilters />      
        <ExpenseList />
    </div>
);
export {ExpenseDashoboardPage as default}