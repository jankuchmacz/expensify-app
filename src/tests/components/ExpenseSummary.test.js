import { shallow } from 'enzyme';
import React from 'react';
import expenses from '../fixtures/expenses';
import {ExpenseSummary} from '../../components/ExpenseSummary';
import getExpensesTotal from '../../selectors/expenses-total';

test('should render correct summary for 1 expense', ()=> {
    const wrapper = shallow(<ExpenseSummary expenses={[expenses[0]]} expensesSum={getExpensesTotal([expenses[0]])} />)
    expect(wrapper).toMatchSnapshot();
})
test('should render correct summary for many expenses', ()=> {
    const wrapper = shallow(<ExpenseSummary expenses={expenses} expensesSum={getExpensesTotal(expenses)} />)
    expect(wrapper).toMatchSnapshot();
})
