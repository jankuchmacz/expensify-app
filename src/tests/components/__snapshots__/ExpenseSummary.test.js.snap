// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correct summary for 1 expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
         
         
      </span>
       
      expenses
       totalling 
      <span>
        $1.95
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render correct summary for many expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
         
         
      </span>
       
      expenses
       totalling 
      <span>
        $1,141.95
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
